File: tests/windows/serve_encoding_probe.py
"""M0 diagnostic: does `dgc serve` keep its NDJSON stdout valid UTF-8 when the interpreter's
stdout encoding is the Windows ANSI code page (cp1252)?

On Windows a piped stdout uses ``locale.getpreferredencoding()`` (cp1252 on en-US) unless UTF-8
mode is on, and ``dgc/`` never reconfigures stdout. This probe drives a real ``dgc serve`` child
(through the SDK, which speaks the protocol) against a loopback mock model whose answer is
``héllo → 你好 🍑``, once with the child forced to cp1252 and once to utf-8, and reports whether
the model text arrived intact. PYTHONIOENCODING is honoured on every OS, so this reproduces off
Windows too; the Windows runner is where it counts for the release.

    DGC_PYTHON=<python with the dgc CLI> python tests/windows/serve_encoding_probe.py

Exit code is always 0: this is a diagnostic, and the workflow leg is non-blocking. The findings
are the ``M0-RESULT`` lines.
"""
from __future__ import annotations

import json
import os
import sys
import tempfile
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

HERE = Path(__file__).resolve().parent
ROOT = HERE.parents[1]
# Import the SDK from source unless a wheel is being tested (DGC_SDK_TEST_INSTALLED=1).
if os.environ.get("DGC_SDK_TEST_INSTALLED") != "1":
    sys.path[:0] = [str(ROOT / "sdk" / "python"), str(ROOT)]

from dgc_sdk import DGC, DGCError  # noqa: E402

ANSWER = "héllo → 你好 🍑"        # é encodes in cp1252; → 你好 🍑 do not
MARKERS = ("héllo", "你好", "🍑")


def _sse(delta: dict, finish: str | None = None) -> str:
    return "data: " + json.dumps({"id": "mock", "object": "chat.completion.chunk",
                                  "choices": [{"index": 0, "delta": delta,
                                               "finish_reason": finish}]}) + "\n\n"


class _Model(BaseHTTPRequestHandler):
    def log_message(self, *_a):  # quiet
        pass

    def do_GET(self):
        body = json.dumps({"data": [{"id": "sdk-model"}]}).encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_POST(self):
        length = int(self.headers.get("Content-Length") or 0)
        self.rfile.read(length)
        payload = (_sse({"content": ANSWER}) + _sse({}, finish="stop") + "data: [DONE]\n\n")
        data = payload.encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.end_headers()
        self.wfile.write(data)


def _start_model():
    server = ThreadingHTTPServer(("127.0.0.1", 0), _Model)
    threading.Thread(target=server.serve_forever, daemon=True).start()
    return server, f"http://127.0.0.1:{server.server_address[1]}/v1"


def _run(base_url: str, encoding: str) -> dict:
    """One serve run with the child's stdout encoding forced to ``encoding``."""
    state = Path(tempfile.mkdtemp(prefix="dgc-enc-state-"))
    work = Path(tempfile.mkdtemp(prefix="dgc-enc-work-"))
    (work / "README.md").write_text("hi\n", encoding="utf-8")
    outcome = {"encoding": encoding, "status": None, "final_text": None,
               "markers_present": None, "error": None}
    try:
        with DGC(state_dir=state, model="sdk-model", base_url=base_url, api_key="sk-local",
                 extra_env={"PYTHONIOENCODING": encoding,
                            "PYTHONUTF8": "1" if encoding.lower().startswith("utf") else "0"}) as dgc:
            session = dgc.session(cwd=work, permissions={"mode": "auto", "unhandled": "deny"})
            result = session.run("Say the greeting exactly.", timeout=120)
            outcome["status"] = result.status
            outcome["final_text"] = result.final_text
            outcome["markers_present"] = sorted(m for m in MARKERS if m in (result.final_text or ""))
    except DGCError as exc:
        outcome["error"] = f"{type(exc).__name__}: {exc}"
    except Exception as exc:  # noqa: BLE001 - a diagnostic records every failure shape
        outcome["error"] = f"{type(exc).__name__}: {exc}"
    return outcome


def main() -> int:
    server, base_url = _start_model()
    try:
        results = [_run(base_url, "utf-8"), _run(base_url, "cp1252")]
    finally:
        server.shutdown()
        server.server_close()

    for r in results:
        intact = r["markers_present"] == sorted(MARKERS) and r["error"] is None
        summary = (f"M0-RESULT windows-encoding platform={sys.platform} "
                   f"enc={r['encoding']} status={r['status']} "
                   f"non_ascii_intact={'YES' if intact else 'NO'} "
                   f"markers={','.join(r['markers_present'] or [])!r} "
                   f"error={r['error']!r}")
        print(summary, flush=True)
        print(f"    final_text={r['final_text']!r}", flush=True)
    print(json.dumps({"probe": "windows-encoding", "platform": sys.platform,
                      "results": results}, ensure_ascii=True), flush=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/windows/serve_encoding_probe.mjs
/**
 * M0 diagnostic (Node side of serve_encoding_probe.py): what does the Node SDK make of a
 * `dgc serve` whose stdout is not UTF-8? The Node transport decodes stdout leniently, so a bad
 * byte is not an error: it becomes U+FFFD, and a frame serve could not encode at all is dropped.
 * The model answers "héllo → 你好 🍑" (é is cp1252-encodable; → 你好 🍑 are not).
 *
 *   DGC_PYTHON=<python with the dgc CLI> node tests/windows/serve_encoding_probe.mjs <sdk dist/index.js>
 *
 * Variants: utf-8 and cp1252 (forced through PYTHONIOENCODING) and native (no override: the
 * interpreter's own default, i.e. what a Windows user gets). Always exits 0; the findings are the
 * ASCII-only `M0-RESULT` lines.
 */
import { createServer } from "node:http";
import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { pathToFileURL } from "node:url";

const entry = resolve(process.argv[2] || "");
const PY = process.env.DGC_PYTHON;
if (!PY) throw new Error("set DGC_PYTHON to the python that has the dgc CLI installed");
const { DGC } = await import(pathToFileURL(entry).href);

const ANSWER = "héllo → 你好 🍑";
const MARKERS = ["héllo", "你好", "🍑"];
const ascii = (value) => JSON.stringify(value).replace(/[^\x20-\x7e]/g,
  (ch) => "\\u" + ch.charCodeAt(0).toString(16).padStart(4, "0"));
const sse = (delta, finish = null) => "data: " + JSON.stringify({ id: "mock", object: "chat.completion.chunk",
  choices: [{ index: 0, delta, finish_reason: finish }] }) + "\n\n";
const server = createServer((req, res) => {
  if (req.method === "GET") {
    res.writeHead(200, { "Content-Type": "application/json" });
    res.end(JSON.stringify({ data: [{ id: "sdk-model" }] }));
    return;
  }
  req.on("data", () => {});
  req.on("end", () => {
    res.writeHead(200, { "Content-Type": "text/event-stream" });
    res.end(sse({ content: ANSWER }) + sse({}, "stop") + "data: [DONE]\n\n");
  });
});
await new Promise((ok) => server.listen(0, "127.0.0.1", ok));
const baseUrl = `http://127.0.0.1:${server.address().port}/v1`;
const work = mkdtempSync(join(tmpdir(), "dgc-enc-work-"));
writeFileSync(join(work, "README.md"), "hi\n");

for (const variant of ["utf-8", "cp1252", "native"]) {
  const extraEnv = variant === "native" ? {} : { PYTHONIOENCODING: variant };
  const dgc = new DGC({ stateDir: mkdtempSync(join(tmpdir(), "dgc-enc-state-")), model: "sdk-model", baseUrl,
    apiKey: "sk-local", runtime: [PY, "-m", "dgc", "serve"], extraEnv });
  let line;
  try {
    const session = await dgc.session({ cwd: work, permissions: { mode: "auto", unhandled: "deny" } });
    const result = await session.run("Say the greeting exactly.", { timeoutMs: 120_000 });
    const present = MARKERS.filter((m) => (result.finalText || "").includes(m));
    const intact = result.finalText === ANSWER;
    line = `status=${result.status} non_ascii_intact=${intact ? "YES" : "NO"} ` +
      `markers=${ascii(present.join(","))} final_text=${ascii(result.finalText)} error=None`;
  } catch (error) {
    line = `status=None non_ascii_intact=NO markers="" final_text=None ` +
      `error=${ascii(`${error?.name}: ${String(error?.message || error).slice(0, 200)}`)}`;
  } finally {
    try { await dgc.close(); } catch { /* recorded above if it mattered */ }
  }
  console.log(`M0-RESULT windows-encoding-node platform=${process.platform} enc=${variant} ${line}`);
}
server.close();
process.exit(0);
